src/auth/webhook-secret.guard.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import * as crypto from 'crypto';
+import { Request } from 'express';
+import { Env } from 'src/env';
+
+@Injectable()
+export class WebhookSecretGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request: Request = context.switchToHttp().getRequest();
+    const signature = request.headers['x-hub-signature-256'] as string;
+    const secret = Env.SECRET.WEBHOOK;
+    const payload = JSON.stringify(request.body);
+
+    if (!signature) {
+      throw new UnauthorizedException('Missing signature');
+    }
+
+    const hmac = crypto.createHmac('sha256', secret);
+    const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+
+    if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
+      throw new UnauthorizedException('Invalid signature');
+    }
+
+    return true;
+  }
```

Comment:
PullProofAI

- The `WebhookSecretGuard` class looks well-implemented for validating webhook signatures using HMAC.
- Consider adding error handling for potential exceptions that may occur during the HMAC calculation process.
- Ensure proper validation and sanitization of the `request.body` before converting it to a JSON string to prevent any potential security vulnerabilities.
- It's recommended to provide more specific error messages to aid in debugging and troubleshooting.

```diff
+ // Add error handling for potential exceptions
+ try {
+     const hmac = crypto.createHmac('sha256', secret);
+     const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+ } catch (error) {
+     throw new UnauthorizedException('Error calculating HMAC');
+ }

+ // Validate and sanitize request body before JSON conversion
+ // Example: Ensure request body is not null or undefined
+ if (!request.body) {
+     throw new UnauthorizedException('Invalid request body');
+ }

+ // Provide more specific error messages for better debugging
+ throw new UnauthorizedException('Signature verification failed');
```

This feedback aims to enhance error handling, security, and maintainability aspects of the `WebhookSecretGuard` implementation.


<!-- This is an auto-generated comment by PullProofAI -->